cpp/src/parquet/encryption/properties_driven_crypto_factory.h

```diff
@@ -0,0 +1,205 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "parquet/encryption/encryption.h"
+#include "parquet/encryption/file_key_wrapper.h"
+#include "parquet/encryption/key_toolkit.h"
+#include "parquet/encryption/kms_client_factory.h"
+#include "parquet/platform.h"
+
+namespace parquet {
+namespace encryption {
+
+static constexpr ParquetCipher::type kDefaultEncryptionAlgorithm =
+    ParquetCipher::AES_GCM_V1;
+static constexpr bool kDefaultPlaintextFooter = false;
+static constexpr bool kDefaultDoubleWrapping = true;
+static constexpr bool kDefaultWrapLocally = false;
+static constexpr uint64_t kDefaultCacheLifetimeSeconds = 600;  // 10 minutes
+static constexpr bool kDefaultInternalKeyMaterial = true;
+static constexpr bool kDefaultUniformEncryption = false;
+static constexpr int32_t kDefaultDataKeyLengthBits = 128;
+
+class PARQUET_EXPORT EncryptionConfiguration {
+ public:
+  class PARQUET_EXPORT Builder {
+   public:
+    /// footer_key: ID of the master key for footer encryption/signing
+    explicit Builder(const std::string& footer_key)
+        : footer_key_(footer_key),
+          encryption_algorithm_(kDefaultEncryptionAlgorithm),
+          plaintext_footer_(kDefaultPlaintextFooter),
+          double_wrapping_(kDefaultDoubleWrapping),
+          cache_lifetime_seconds_(kDefaultCacheLifetimeSeconds),
+          internal_key_material_(kDefaultInternalKeyMaterial),
+          uniform_encryption_(kDefaultUniformEncryption),
+          data_key_length_bits_(kDefaultDataKeyLengthBits) {}
+
+    /// List of columns to encrypt, with master key IDs (see HIVE-21848).
+    /// Format: "masterKeyID:colName,colName;masterKeyID:colName..."
+    /// Either
+    /// column_keys(const std::string&)
+    /// or
+    /// uniform_encryption()
+    /// must be called. If none are called, or if both are called, an exception will be
+    /// thrown.
+    Builder* column_keys(const std::string& column_keys);
+
+    /// encrypt footer and columns with the same encryption key
+    Builder* uniform_encryption();
+
+    /// Parquet encryption algorithm. Can be "AES_GCM_V1" (default), or "AES_GCM_CTR_V1".
+    Builder* encryption_algorithm(ParquetCipher::type algo);
+
+    /// Write files with plaintext footer
```

Comment:
It would be useful to spell out the default values in the public API documentation.  E.g.,
". The default is false - files are written with encrypted footer".